src/DynamoCore/Nodes/NodeCategories.cs

```diff
@@ -430,28 +430,150 @@ internal static string MakeAbsolutePath(string basePath, string relativePath)
-                var curr = original[i];
-                var prev = original[i - 1];
-                if ((Char.IsUpper(curr) || curr.Equals('(')) &&
-                    ((prev != ' ') && (!Char.IsUpper(prev))))
+                var word = words.ElementAt(i);
+                if ((currentWord.Length + word.Length + 1) <= maxCharacters)
                 {
-                    newText.Append(" ");
+                    currentWord.Append(" ");
+                    currentWord.Append(word);
                 }
-                newText.Append(original[i]);
+                else
+                {
+                    result_words.Add(currentWord.ToString());
+                    currentWord.Clear();
+                    currentWord.Append(word);
+                }
+            }
+            // Add last word.
+            if (currentWord.Length != 0)
+                result_words.Add(currentWord.ToString());
+
+            return result_words;
+        }
+
+        /// <summary>
+        /// Reduces the number of rows, based on the entries inside rows parameter.
+        /// E.g. rows = { "Insert", "Day", "Of", "Week", "Here" }, maxRows == 3
+        /// Result { "Insert", "Day", "Of Week Here" }
+        /// </summary>
+        /// <param name="rows">Incoming rows</param>
+        /// <param name="maxRows">Max number of rows</param>
+        internal static IEnumerable<string> ReduceRowCount(List<string> rows, int maxRows)
+        {
+            if (rows == null || maxRows <= 0)
+                throw new ArgumentException();
+            if (rows.Count() == 0 || rows.Count() <= maxRows)
+                return rows;
+
+            while (rows.Count() != maxRows)
+            {
+                int penultimateIndex = rows.Count() - 2;
+                var penultimateRow = rows.ElementAt(penultimateIndex);
+                var lastRow = rows.Last();
+
+                string mergedRow = String.Concat(penultimateRow, " ", lastRow);
+                
+                // Remove 2 old rows and add 1 new.
+                rows.RemoveRange(penultimateIndex, 2);
+                rows.Add(mergedRow);
             }
-            return newText.ToString();
+
+            return rows;
+        }
+
+        /// <summary>
+        /// Truncate each entry in the given "rows" to a maximum of "maxCharacters".
+        /// For examples, given that "maxCharacters" equals to "8":
+        /// 
+        ///     { "Surface", "Analysis Data" } => { "Surface", "..s Data" }
+        ///     { "By", "Geometry", "Coordinate", "System" } => { "By", "Geometry", "Coordi..", "System" }
+        ///     { "By Geometry", "Coordinate System" } => { "By Geo..", "..System" }
+        /// </summary>
+        /// <param name="rows">Incomming rows</param>
+        /// <param name="maxCharacters">Max number characters per row</param>
+        internal static IEnumerable<string> TruncateRows(IEnumerable<string> rows, int maxCharacters)
```

Comment:
I think the method is overly complicated, can you try something like this (I typed it in Notepad++, so might contains errors):

```csharp
var maxAfterTruncate = maxCharacters - twoDotsLength;
var results = new List<string>();
foreach (var row in rows)
{
    if (row.Length <= maxCharacters) {
        results.Add(row);
        continue;
    }
    
    if (row != lastRow)
    {
        // Rows other than the last get dots appended to the end.
        results.Add(row.SubString(0, maxAfterTruncate) + Configurations.TwoDots);
    }
    else
    {
        // The final row gets the dots added to the front. 
        var offset = row.Length - maxAfterTruncate;
        results.Add(Configurations.TwoDots + row.SubString(offset));
    }
}
```